akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -1393,33 +1500,104 @@ private[cluster] final class FirstSeedNodeProcess(seedNodes: immutable.IndexedSe
 
   if (seedNodes.isEmpty || seedNodes.head == selfAddress)
     throw new IllegalArgumentException("Join seed node should not be done")
 
-  context.setReceiveTimeout(Cluster(context.system).settings.SeedNodeTimeout)
+  context.setReceiveTimeout(SeedNodeTimeout)
 
   var attempt = 0
 
+  // all seed nodes, except this one
+  val otherSeedNodes = seedNodes.toSet - selfAddress
+
   override def preStart(): Unit = self ! JoinSeedNode
 
   def receive = {
     case JoinSeedNode ⇒
+      val requiredNonSensitiveKeys = JoinConfigCompatChecker.removeSensitiveKeys(joinConfigCompatChecker.requiredKeys, cluster.settings)
+      // configToValidate only contains the keys that are required according to JoinConfigCompatChecker on this node
+      val configToValidate = JoinConfigCompatChecker.filterWithKeys(requiredNonSensitiveKeys, context.system.settings.config)
       // send InitJoin to all seed nodes (except myself)
       attempt += 1
-      seedNodes.collect {
-        case a if a != selfAddress ⇒ context.actorSelection(context.parent.path.toStringWithAddress(a))
-      } foreach { _ ! InitJoin }
-    case InitJoinAck(address) ⇒
-      // first InitJoinAck reply
-      context.parent ! JoinTo(address)
-      context.become(done)
+      otherSeedNodes.foreach { a ⇒ context.actorSelection(context.parent.path.toStringWithAddress(a)) ! InitJoin(configToValidate) }
+
+    case InitJoinAck(address, CompatibleConfig(clusterConfig)) ⇒
+      log.info("Received InitJoinAck message from [{}] to [{}]", sender(), selfAddress)
+      // validates config coming from cluster against this node config
+      joinConfigCompatChecker.check(clusterConfig, context.system.settings.config) match {
+        case Valid ⇒
+          // first InitJoinAck reply
+          context.parent ! JoinTo(address)
+          context.become(done)
+
+        case Invalid(messages) if ByPassConfigCompatCheck ⇒
+          log.warning("Cluster validated this node config, but sent back incompatible settings: {}. " +
+            "Join will be performed because compatibility check is configured to not be enforced.", messages.mkString(", "))
+          context.parent ! JoinTo(address)
+          context.become(done)
+
+        case Invalid(messages) ⇒
+          log.error("Cluster validated this node config, but sent back incompatible settings: {}", messages.mkString(", "))
+          log.error(
+            "It's recommended to perform a full cluster shutdown in order to deploy this new version." +
+              "If a cluster shutdown isn't an option, you may want to disable this protection by setting " +
+              "'akka.cluster.configuration-compatibility-check.enforce-on-join = off'. " +
+              "Note that disabling it will allow the formation of a cluster with nodes having incompatible configuration settings. " +
+              "This node will be shutdown!")
+          CoordinatedShutdown(context.system).run(IncompatibleConfigurationDetected)
```

Comment:
do we need to stop this actor also, since it may take a while until the shutdown completes?